FHEM/00_SIGNALduino.pm

```diff
@@ -1869,7 +1873,11 @@ sub SIGNALduino_Split_Message($$)
 			$_ =~ s/R=//;
 			$rssi = $_ ;
 			Debug "$name: extracted RSSI $rssi \n" if ($debug);
-			$ret{rssi} = $rssi;
+			if ($rssi == 0) {
+				$ret{rssi} = undef;
```

Comment:
Daa würde ich mir gerne noch Mal genauer ansehen wieso wir den Wert auf undef setzen und nicht alternativ einfach den Key rssi nicht setzen, wenn es keinen Wert gibt.

Reply:
In dem PR setzen wir den Wert auf undef, weil sonst bei einem LowBudget Empfänger auch das Internal geschrieben wird mit dem Wert -74 obwohl der keinen RSSI besitzt.

Sobald wir den RSSI mit ins Internal schreiben, Befüllen wir auch „sinnloser Weise“ das Log von solch einem Empfänger mit einem Wert den er nicht besitzt.

Cc1101 hat RSSI somit ist es richtig im Internal und im Logfile.

Reply:
Der RSSI wird auf undef gesetzt, weil die folgenden Routinen auf defined prüfen.
Sinnvoller wäre es aber sicher, wenn der SIGNALduino bzw. SIGNALESP den RSSI gar nicht erst übergibt.

Reply:
Dies ist ein bug in der firmware V 3.3.1 RC und da sollte er auch behoben werden.
In der V 3.3.1-dev müsste es noch funktionieren 
Ich hatte es mal eingebaut aber inzwischen wurde es wieder entfernt
https://github.com/RFD-FHEM/SIGNALDuino/commit/79d97fdc0d713364612d2ac69c2439ed81b00321

Reply:
@HomeAutoUser 

Wäre es so nicht etwas besser? Wir setzen den Key {rssi} nur, wenn der Wert > 0 ist. Und wenn der Key im hash nicht existiert, dann haben wir keinen RSSI Wert.

```
 FHEM/00_SIGNALduino.pm | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/FHEM/00_SIGNALduino.pm b/FHEM/00_SIGNALduino.pm
index 087f748..ba05fc0 100644
--- a/FHEM/00_SIGNALduino.pm
+++ b/FHEM/00_SIGNALduino.pm
@@ -1869,7 +1869,7 @@ sub SIGNALduino_Split_Message($$)
 			$_ =~ s/R=//;
 			$rssi = $_ ;
 			Debug "$name: extracted RSSI $rssi \n" if ($debug);
-			$ret{rssi} = $rssi;
+			$ret{rssi} = $rssi if ($rssi > 0);
 		}  else {
 			Debug "$name: unknown Message part $_" if ($debug);;
 		}
@@ -1986,13 +1986,12 @@ SIGNALduino_Parse_MS($$$$%)
 	my $protocolid;
 	my $syncidx=$msg_parts{syncidx};			
 	my $clockidx=$msg_parts{clockidx};				
-	my $rawRssi=$msg_parts{rssi};
 	my $protocol=undef;
 	my $rawData=$msg_parts{rawData};
 	my %patternList;
-	my $rssi;
-	if (defined($rawRssi)) {
-		$rssi = ($rawRssi>=128 ? (($rawRssi-256)/2-74) : ($rawRssi/2-74)); # todo: passt dies so? habe ich vom 00_cul.pm
+	my $rssi=undef;
+	if (exists{$msg_parts{rssi}} && defined({$msg_parts{rssi}})  {
+		$rssi = ($msg_parts{rssi}>=128 ? (($msg_parts{rssi}-256)/2-74) : ($msg_parts{rssi}/2-74)); # todo: passt dies so? habe ich vom 00_cul.pm
 	}
     #$patternList{$_} = $msg_parts{rawData}{$_] for keys %msg_parts{rawData};
 

```

Reply:
Ich hab mal getestet, in dieser firmware ist der bug noch nicht drin
`V 3.3.1-dev SIGNALduino - compiled at Jan 3 2017 23:59:32`

Reply:
Wie äußert sich der Big? Ich habe etwas aufgedeckt, sobald R=0 wird immer der Wert -74 eingetragen. Somit ergibt sich ein Kamm in der RSSI Darstelllung. In meinem Fall kommt der Sensor mit ca.-94dB regelmäßig wo R=216 ist und sobald mal R=0 kommt, wird -74 geschrieben.

Reply:
@Ralf9 
Ich habe das jetzt mal versucht, in die aktuelle Version einzubinden:
signalDecoder.cpp (sieht mittlerweile etwas anders aus):

#ifdef CMP_CC1101
          n = sprintf(buf, ";C%X;S%X;R%X;", clock, sync, rssiValue);
#else
          n = sprintf(buf, ";C%X;S%X;", clock, sync);
#endif
					SDC_WRITE((const uint8_t *)buf, n);

Funktioniert zwar im Prinzip, aber das "#define CMP_CC1101" aus der "SIGNALDuino.ino" wird dort nicht ausgewertet. Es wird also immer als "nicht definiert" verwendet.
Gibt es da eine Möglichkeit, das "#define ..." der Routine zu übergeben?
Man könnte zwar auch die Variable "hasCC1101" auswerten, aber das verbrät dann ziemlich viel Speicher. Die RSSi-Ausgabe wird ja 5x verwendet.

Reply:
Ich kenne nur die Möglichkeit das `#define CMP_CC1101` in die SIGNALDuino.h zu schreiben und damit funktioniert es.

Reply:
wir könnten auf den Wert von  rssiValue an dieser Stelle prüfen. Das wird ja auf NULL über die callback gesetzt, wenn kein CC1101 vorhanden ist.

Reply:
`Gibt es da eine Möglichkeit, das "#define ..." der Routine zu übergeben?`

Ja, die Möglichkeit bedeutet, dass alle #defines in eine eigene header Datei kommen config.h z.B. diese wird dann überall eingebunden. Eventuell muss ich das ohnehin machen, da ich gerade noch nebenbei versuche SIGNALduino und SIGNALESP zusammen zu führen.

Die 2. Option besteht darin, beim compilieren die Option mit anzugeben. "compiler.cpp.extra_flags". 
So mache ich das heute bereits um die verschiedenen Board Typen zu compilieren.

Reply:
Ich habe die Konfiguration jetzt in einer extra Datei, ähnlich wie du es jetzt eingbaut hast. Müsste ich dann nur in die neue Version übernehmen. Hat den großen Vorteil, das Einstellungen zentral in einer Datei vorgenommen werden können.

Reply:
Ja, es ist aber schlecht in einer Bibliothek ein include auf eine Datei aus einem Projekt vorzunehmen.
Die Bibliothek funktioniert dann ja nicht so wie sie sollte.

Daher hatte ich das bisher über callback und co gelöst.

Reply:
Mhmm, verstehe ich jetzt nicht ganz. Du hast doch die erforderlichen Defines jetzt sowieso schon in der compile_config.h stehen. Da ist es doch ein leichtes Unterfangen, das jetzt abzufragen:

#ifdef CMP_CC1101
n = sprintf(buf, ";C%X;S%X;R%X;", clock, sync, rssiValue);
#else
n = sprintf(buf, ";C%X;S%X;", clock, sync);
#endif

Reply:
compile_config.h ist Bestandteil des Hauptprogrammes

signaldecoder.h ist Bestandteil einer Bibliothek die prinzipiell nichts mit dem Hauptprogramm zu tun haben muss. Wenn Du dir mal z.B. das unit Test Programm ansiehst, dann weisst Du was ich meine. Dort verwende ich die gleiche Bibliothek aber natürlich ein anderes Hauptprogramm.

Wenn die Bibliothek compiliert wird, dann wird compile_config.h nicht eingebunden.

Reply:
Und was spricht gegen ein #include "compile_config.h" in der "signalDecoder.cpp"? Der verwendete Prozessor muss doch sowieso immer definiert werden.

Reply:
In einer Bibliothek bindet man keinen Quellcode vom Hauptprojekt ein.

Wieso nicht einfach prüfen ob der Funktionszeiger auf NULL zeigt?

Reply:
Hab ich keine Ahnung, wie das geht.

Reply:
Ich baue das ein.